proxy/src/transparency/tcp.rs

```diff
@@ -71,14 +73,177 @@ impl Proxy {
         let fut = connect.connect()
             .map_err(|e| debug!("tcp connect error: {:?}", e))
             .and_then(move |tcp_out| {
-                let (in_r, in_w) = tcp_in.split();
-                let (out_r, out_w) = tcp_out.split();
-
-                copy(in_r, out_w)
-                    .join(copy(out_r, in_w))
-                    .map(|_| ())
+                Duplex::new(tcp_in, tcp_out)
                     .map_err(|e| debug!("tcp error: {}", e))
             });
         Box::new(fut)
     }
 }
+
+/// A future piping data bi-directionally to In and Out.
+struct Duplex<In, Out> {
+    in_buf: CopyBuf,
+    in_eof: bool,
+    in_io: In,
+    in_shutdown: bool,
+    out_buf: CopyBuf,
+    out_eof: bool,
+    out_io: Out,
+    out_shutdown: bool,
+}
+
+/// A buffer used to copy bytes from one IO to another.
+///
+/// Keeps read and write positions.
+struct CopyBuf {
+    // In linkerd-tcp, a shared buffer is used to start, and an allocation is
+    // only made if NotReady is found trying to flush the buffer. We could
+    // consider making the same optimization here.
+    buf: Box<[u8]>,
+    read_pos: usize,
+    write_pos: usize,
+}
+
+impl<In, Out> Duplex<In, Out>
+where
+    In: AsyncRead + AsyncWrite,
+    Out: AsyncRead + AsyncWrite,
+{
+    fn new(in_io: In, out_io: Out) -> Self {
+        Duplex {
+            in_buf: CopyBuf::new(),
+            in_eof: false,
+            in_io,
+            in_shutdown: false,
+            out_buf: CopyBuf::new(),
+            out_eof: false,
+            out_io,
+            out_shutdown: false,
+        }
+    }
+
+    /// Returns true if the Duplex is completely done.
+    ///
+    /// The future may finish earlier if one side has an error.
+    fn is_done(&self) -> bool {
+        self.in_eof
+            && !self.in_buf.has_remaining()
+            && self.out_eof
+            && !self.out_buf.has_remaining()
```

Comment:
If we refactor this to remove the duplication like I suggest above then I think this would become something like `self.in.is_done() && self.out.is_done()`.